// rc-scroll-anim/lib/ScrollOverPack.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _objectWithoutProperties2 = require('babel-runtime/helpers/objectWithoutProperties');

var _objectWithoutProperties3 = _interopRequireDefault(_objectWithoutProperties2);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _EventDispatcher = require('./EventDispatcher');

var _EventDispatcher2 = _interopRequireDefault(_EventDispatcher);

var _ScrollElement2 = require('./ScrollElement');

var _ScrollElement3 = _interopRequireDefault(_ScrollElement2);

var _util = require('./util');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function noop() {}

var ScrollOverPack = function (_ScrollElement) {
  (0, _inherits3["default"])(ScrollOverPack, _ScrollElement);

  function ScrollOverPack(props) {
    (0, _classCallCheck3["default"])(this, ScrollOverPack);

    var _this = (0, _possibleConstructorReturn3["default"])(this, _ScrollElement.call(this, props));

    _this.scrollEventListener = function (e) {
      _this.getParam(e);
      if (_this.enter) {
        if (!_this.state.show) {
          _this.setState({
            show: true
          });
        }
        if (!_this.props.always) {
          _EventDispatcher2["default"].removeEventListener(_this.eventType, _this.scrollEventListener);
        }
      }
      var bottomLeave = _this.elementShowHeight < _this.playHeight;
      // 设置往上时的出场点...
      var topLeave = _this.props.replay ? _this.elementShowHeight > _this.clientHeight + _this.leavePlayHeight : null;
      if (topLeave || bottomLeave) {
        if (_this.state.show) {
          _this.setState({
            show: false
          });
        }
      }
    };

    _this.children = (0, _util.toArrayChildren)(_this.props.children);
    _this.oneEnter = false;
    _this.enter = false;
    _this.state = {
      show: false,
      children: (0, _util.toArrayChildren)(_this.props.children)
    };
    return _this;
  }

  ScrollOverPack.prototype.render = function render() {
    var _this2 = this;

    var placeholderProps = (0, _objectWithoutProperties3["default"])(this.props, []);

    ['playScale', 'replay', 'component', 'always', 'scrollEvent', 'hideProps', 'location'].forEach(function (key) {
      return delete placeholderProps[key];
    });
    var childToRender = void 0;
    if (!this.oneEnter && !this.state.show) {
      childToRender = (0, _react.createElement)(this.props.component, (0, _extends3["default"])({}, placeholderProps), null);
      this.oneEnter = true;
    } else {
      if (!this.state.show) {
        this.children = this.children.map(function (item) {
          if (!item || !item.key) {
            return null;
          }
          var element = void 0;
          var hideProps = _this2.props.hideProps[item.key];
          if (hideProps) {
            element = _react2["default"].cloneElement(item, (0, _extends3["default"])({}, hideProps));
            return element;
          }
          element = _react2["default"].cloneElement(item, {}, null);
          return element;
        });
      } else {
        this.children = this.state.children;
      }
      childToRender = (0, _react.createElement)(this.props.component, (0, _extends3["default"])({}, placeholderProps), this.children);
    }
    return childToRender;
  };

  return ScrollOverPack;
}(_ScrollElement3["default"]);

ScrollOverPack.propTypes = {
  component: _react2["default"].PropTypes.string,
  playScale: _react2["default"].PropTypes.any,
  always: _react2["default"].PropTypes.bool,
  scrollEvent: _react2["default"].PropTypes.func,
  children: _react2["default"].PropTypes.any,
  className: _react2["default"].PropTypes.string,
  style: _react2["default"].PropTypes.any,
  replay: _react2["default"].PropTypes.bool,
  onChange: _react2["default"].PropTypes.func,
  hideProps: _react2["default"].PropTypes.object
};

ScrollOverPack.defaultProps = {
  component: 'div',
  playScale: 0.5,
  always: true,
  scrollEvent: noop,
  replay: false,
  onChange: noop,
  hideProps: {}
};

exports["default"] = ScrollOverPack;
module.exports = exports['default'];
